Send NAT probing data to Rappor

kennysong opened the pull request:
Submits the NAT type, NAT-PMP, PCP, and UPnP support data to Rappor.

#1646.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1652)
<!-- Reviewable:end -->

bemasc commented on src/generic_core/metrics.ts:
```diff
@@ -105,7 +143,9 @@ export class DailyMetricsReporter {
   constructor(private metrics_ :Metrics, private storage_ :storage.Storage,
+              private getNetworkInfo_ :Function,
               private onReportCallback_ :Function) {
+
```
No blank line.

jpevarnek commented:
This looks awesome!

Couple small things I had questions about, mostly stylistic, you should take any comments I make as me thinking "Oh, this stands out, I think it might be useful to have a short discussion about it"

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/metrics.ts, line 85 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1YSj9u699lFZHBlck)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L85)):</sup>
Using substring here with arbitrary numbers for the string position is a bit hard to understand with parsing, and will completely break if how we display things to the user changes.  Is there any way you could call a function that returns an object here and deal with changing that into something that makes sense to a human at a different point?

---

<sup>**[src/generic_core/metrics.ts, line 85 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1_wnj3T9J3JXWEmtH)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L85)):</sup>
Is it necessary to store these to the object?  It does not look like we ever look at them except in this function when we have just stored them.

---

<sup>**[src/generic_core/metrics.ts, line 92 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1Yx0QAcRSmyJ3IfGI)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L92)):</sup>
Right now, this reads as:

* Get the network info
* Wait for us to load the metrics data from  disk
* Do work

In our case, it does not make too much of a difference, but these should really not be blocking on each other.  You could do something closer to

* Get the network info and wait for us to load the metrics data from disk
* Do work

```
return Promise.all([ getNetworkInfo(), this.onceLoaded_ ]).then((values) => {
  networkInfo = values[0];
  ...
  return Promise.all(reports);
}).then(this.metricsProvider_.retrieve);
```

I think that would look a bit cleaner (and would get rid of some unnecessary blocking), but it's obviously completely your call

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/metrics.ts, line 168 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1ag2swgXzRx87bLP-)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L168)):</sup>
Why pass a function that will then immediately be called?  Why not just call `this.getNetworkInfo_` before `getReport`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

dborkan commented on src/generic_core/metrics.ts:
```diff
@@ -25,9 +30,16 @@ export class Metrics {
       flag_oneprr: true
     };
+    var stringMetric = {
+      type: 'string', num_bloombits: 8, num_hashes: 2,
```
@willscott @mollyling - can you folks double check these values here?  We recently had an issue regarding an incorrect setting for the bloombits (#1644), want to make sure we won't have similar issues again

willscott replied:
How many expected possibilities are there for these values?

dborkan commented on src/generic_core/metrics.ts:
```diff
@@ -11,13 +10,19 @@ export interface MetricsData {
+  natType :string;
+  pmpSupport :string;  // Either 'Supported' or 'Not supported'
+  pcpSupport :string;
+  upnpSupport :string;
```
This is kinda optional, but I think it would be fine to leave these new fields out of MetricsData.  MetricsData is really just what we read/write to storage.  In the case of success and failure counts this is needed because we want to keep track of the total counts in about one day, even across restarts of uProxy.  But for these new fields, we just query the natType, etc at the time we generate the report, and they never actually need to be saved.

willscott commented on src/generic_core/metrics.ts:
```diff
@@ -25,9 +30,16 @@ export class Metrics {
     };
+    var stringMetric = {
+      type: 'string', num_bloombits: 8, num_hashes: 2,
+      num_cohorts: 64, prob_p: 0.5, prob_q: 0.75, prob_f: 0.5,
```
@bemasc - we should figure out some initial values for these probs that add a bit less noise so we can extract stuff out earlier. No reason to switch the success/failure ones, but may be worth getting some extra fidelity out of these.

kennysong commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/metrics.ts, line 34 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1WbAnZzJJaMHNJtrs-r1-34)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L34)):</sup>
For the `natType` string, there are 4 values. The other three only take on 2 values.

---

<sup>**[src/generic_core/metrics.ts, line 85 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1_wnj3T9J3JXWEmtH)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L85)):</sup>
It's not necessary – I think we could remove them from the `data_` object  and `metricsData` interface as well. However, the code would be less consistent and it would be less clear what metrics were being sent to Rappor (though the 4 new metrics would still be documented near the top in the `metricsProvider_` definition). What do you think?

---

<sup>**[src/generic_core/metrics.ts, line 168 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1ag2swgXzRx87bLP-)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L168)):</sup>
We could call `getNetworkInfo_()` and pass its output as an argument into `getReport()`. Would that be more clean? (It would still be blocking, as noted by @jpevarnek's comment)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

willscott commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/metrics.ts, line 34 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1WbAnZzJJaMHNJtrs-r1-34)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L34)):</sup>
8 bits should be enough, but maybe this should be a 'natMetric', rather than a generic 'stringMetric', since we'll want more bits configured for other strings we report.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
<sup>**[src/generic_core/metrics.ts, line 168 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1ag2swgXzRx87bLP-)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L168)):</sup>
FWIW, I think the cleanliness boost from doing it that way would make up for whatever couple cycles you may lose.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/metrics.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1WbAnZzJJaMHNJtrs-r1-16)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L16)):</sup>
Done.

---

<sup>**[src/generic_core/metrics.ts, line 34 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1WbAnZzJJaMHNJtrs-r1-34)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L34)):</sup>
Done.

---

<sup>**[src/generic_core/metrics.ts, line 85 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1YSj9u699lFZHBlck)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L85)):</sup>
I replaced this with a new function `getNetworkInfoObj()` that returns an object with the NAT probing information, which is more robust.

---

<sup>**[src/generic_core/metrics.ts, line 85 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1_wnj3T9J3JXWEmtH)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L85)):</sup>
Given Daniel's comment (line 16), I went ahead and moved the new metrics out of the object and interface.

---

<sup>**[src/generic_core/metrics.ts, line 92 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1Yx0QAcRSmyJ3IfGI)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L92)):</sup>
I moved the `getNetworkInfo()` call out of `getReport()` to follow Ben's suggestion. It's still blocking but the flow is cleaner now.

---

<sup>**[src/generic_core/metrics.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1WbAnZzJJaMHNJtrs-r1-148)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L148)):</sup>
Done.

---

<sup>**[src/generic_core/metrics.ts, line 168 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js1ag2swgXzRx87bLP-)** ([raw file](https://github.com/uproxy/uproxy/blob/10fa98b3e0db9a70d98edb214f8fb80348a0d80a/src/generic_core/metrics.ts#L168)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
LGTM

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/metrics.ts, line 129 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6246HeBoXv-E7OEpK)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/metrics.ts#L129)):</sup>
Why are you passing this as an argument to the constructor, when it's a static function that's already in scope as `uproxy_core.getNetworkInfoObj`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

bemasc commented:
@dborkan, @jpevarnek: Please check these suggestions to make all this stuff static.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/uproxy_core.ts, line 321 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6B5E8c5L9Y5WUgETF)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L321)):</sup>
This can be static.

---

<sup>**[src/generic_core/uproxy_core.ts, line 323 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6BEWOenf-hyCFWNGP)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L323)):</sup>
Once `natResetTimeout_` is static, this can be static too.

---

<sup>**[src/generic_core/uproxy_core.ts, line 362 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js699Ws9_oxRG61L7Pt)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L362)):</sup>
This can be static.

---

<sup>**[src/generic_core/uproxy_core.ts, line 369 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js69DqoqN0254kcCJ0N)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L369)):</sup>
If `routerIps` is static, this can be static.

---

<sup>**[src/generic_core/uproxy_core.ts, line 385 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6BLEd9_Uwf_CsWbE1)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L385)):</sup>
This can be static.

---

<sup>**[src/generic_core/uproxy_core.ts, line 396 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6BNQNY0c7ahfmy61F)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L396)):</sup>
Then this can be static.

---

<sup>**[src/generic_core/uproxy_core.ts, line 426 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6BSvTR0m09siYbAPM)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/uproxy_core.ts#L426)):</sup>
This too.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
All of that looks fine, I'd recommend putting it off until a later PR though.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

kennysong commented:
Changing these members to static now would let us use `uProxyCore.getNetworkInfo()` in `metrics.ts` without having to pass in the function as an argument, which would be a lot cleaner.

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Ah, okay.  If you want to do it now, go for it!

kennysong commented:
I couldn't actually change this because of a circular dependency (https://github.com/uProxy/uproxy/issues/1660). I added a todo to refactor the code once the issue is fixed.

---

Reviewed 4 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/metrics.ts, line 129 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1652#-Js6246HeBoXv-E7OEpK)** ([raw file](https://github.com/uproxy/uproxy/blob/cdf4ed777b35720e08308d00440e547895c3abb1/src/generic_core/metrics.ts#L129)):</sup>
We can't access `uproxy_core` directly because of circular dependencies. See: https://github.com/uProxy/uproxy/issues/1660

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->

bemasc commented:
:+1: 

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1652)
<!-- Sent from Reviewable.io -->
